[abandoned] Dynamo traces distributed.allreduce

wconstab opened the pull request:
We will need to revisit the dynamo support for traceable collectives based on latest design.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90024
* __->__ #89984

- a new guard type (PROCESS_GROUP_MATCH) enforces critical properties
  of the PG match at runtime, while not exhaustively safeguarding
  other transient state vars of c++ PG impl
- allreduce is inplace-only, and stream is ignored

Based on work from Yifu Wang

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

wconstab committed:
Dynamo traces distributed.allreduce

- a new guard type (PROCESS_GROUP_MATCH) enforces critical properties
  of the PG match at runtime, while not exhaustively safeguarding
  other transient state vars of c++ PG impl
- allreduce is inplace-only, and stream is ignored

Based on work from Yifu Wang

[ghstack-poisoned]

wconstab committed:
Update on "Dynamo traces distributed.allreduce"


- a new guard type (PROCESS_GROUP_MATCH) enforces critical properties
  of the PG match at runtime, while not exhaustively safeguarding
  other transient state vars of c++ PG impl
- allreduce is inplace-only, and stream is ignored

Based on work from Yifu Wang

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Dynamo traces distributed.allreduce"


- a new guard type (PROCESS_GROUP_MATCH) enforces critical properties
  of the PG match at runtime, while not exhaustively safeguarding
  other transient state vars of c++ PG impl
- allreduce is inplace-only, and stream is ignored

Based on work from Yifu Wang

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Dynamo traces distributed.allreduce"


- a new guard type (PROCESS_GROUP_MATCH) enforces critical properties
  of the PG match at runtime, while not exhaustively safeguarding
  other transient state vars of c++ PG impl
- allreduce is inplace-only, and stream is ignored

Based on work from Yifu Wang

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Dynamo traces distributed.allreduce"


- a new guard type (PROCESS_GROUP_MATCH) enforces critical properties
  of the PG match at runtime, while not exhaustively safeguarding
  other transient state vars of c++ PG impl
- allreduce is inplace-only, and stream is ignored

Based on work from Yifu Wang

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab commented on torch/_dynamo/utils.py:
```diff
@@ -1038,6 +1051,11 @@ def get_fake_value(node, tx):
     def visit(n: torch.fx.Node):
         return n.meta["example_value"]
 
+    if node.op == "call_function" and node.target in DYNAMO_SUPPORTED_COLLECTIVES:
```
@voznesenskym the primary reason for this special case is that allreduce doesn't work with fake/meta tensors, so we can't actually run it.

but the secondary issue i ran into and also conveniently worked around is that the line below (map_arg, ..., visit) blows up on allreduce, since its processgroup input is an fx.node with no meta['example_input'].

I messed around with how to properly add that briefly and wasn't sure what the invariants are or whether a processgroup even should have its meta filled out.  The logical value there would just be the processgroup, afaiu.

wconstab replied:
i don't mind taking another pass myself, so i can learn. i'm working on meta for allreduce first

voznesenskym commented on torch/_dynamo/guards.py:
```diff
@@ -470,6 +470,21 @@ def TENSOR_MATCH(self, guard: Guard):
+
+        code = list()
+        # type_id encompasses NCCL vs Gloo
+        code.append(f"___check_type_id({ref}, {self.id_ref(t)})")
```
use the TYPE_MATCH guard directly please for this line

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -208,6 +210,17 @@ def _wrap(self, value):
+            # process group which is required for passing it to __compiled_fn.
+            self.tx.output.graphargs.append(GraphArg(self.get_source(), value, False))
+            return ProcessGroupVariable(
+                proxy=self.tx.output.create_graph_input(
+                    re.sub(r"[^a-zA-Z0-9]+", "_", self.name), type(value)
+                ),
```
Hmm, we generally limit the `output.create_graph_input` calls to wrappers, as this is not really a graph input, per se. 

One issue is that this will make a `placeholder` aka fx graph input arg, which is probably not what you want, right?

wconstab replied:
what are my options? bake the PG into the graphmodule instead? or something else?

voznesenskym replied:
That depends, is this primarily for export or eager? For eager, i'd say installing as a global might not be awful. For export, we need to think of a strategy where a placeholder might end up being what you want.

yifuwang replied:
I think baking PGs into graph could be a viable option though not as-is.

Let's assume that:
- we can "decay" a process group into a tuple of integers (e.g. membership in terms of global ranks)
- we have a new set of collective APIs that understands the integer representation of groups (let's make these functional while we are at it)

With the above, I wonder if it's possible/appropriate for dynamo to:
- translate a group to the integer representation
- translate the existing collectives (e.g. dist.all_reduce) to the new calls

cc: @aazzolini @mrshenli @wanchaol

mrshenli replied:
> we can "decay" a process group into a tuple of integers (e.g. membership in terms of global ranks)

Do we need to cover the cases where there are two PGs with the same set of ranks? For that, we might need an additional identifier than just member ranks?

H-Huang commented on torch/_dynamo/utils.py:
```diff
@@ -146,6 +147,18 @@ def fmt_fn(values, item_fn=lambda x: x):
+}
+
+DYNAMO_SUPPORTED_PROCESS_GROUPS = {
+    torch.distributed.ProcessGroupGloo,
+}
+if hasattr(torch.distributed, "ProcessGroupNCCL"):
+    DYNAMO_SUPPORTED_PROCESS_GROUPS.add(torch.distributed.ProcessGroupNCCL)
+
```
Just wanted to mention this feature is being worked on (https://github.com/pytorch/pytorch/issues/86225), with the main PR https://github.com/pytorch/pytorch/pull/88330 still in progress. This will change how ProcessGroups are initialized. `init_process_group()` will not longer set the default process group to be of type `torch.distributed.ProcessGroupNCCL` or `torch.distrbuted.ProcessGroupGloo` etc., but rather just `torch.distributed.ProcessGroup`. When a collective is called such as `dist.allreduce()` it will use `torch.distributed.ProcessGroup` internally route it to a C++ instance of `ProcessGroupGloo` / `ProcessGroupNCCL` based on the collective arguments. This change may cause an issue with PG matching.

wconstab replied:
oh, thanks! i think that will only make things easier - i can just simplify this part to only support ProccessGroup - right?

H-Huang replied:
I think so, these changes arent specific to only gloo and nccl backend right? Will other lesser used ProcessGroup backends be supported such as ProcessGroupMPI, ProcessGroupUCC, or custom backends (e.g. XLA). If these changes are agnostic to that, then just `ProcessGroup` works!

yifuwang commented on torch/_dynamo/guards.py:
```diff
@@ -470,6 +470,21 @@ def TENSOR_MATCH(self, guard: Guard):
+        code = list()
+        # type_id encompasses NCCL vs Gloo
+        code.append(f"___check_type_id({ref}, {self.id_ref(t)})")
+        # size/rank appear to be the only other critical attributes
```
I think we also need to guard group membership in terms of the global ranks. A counter example where guarding (world_size, rank) is not robust:

```python
def main() -> None:
    dist.init_process_group("gloo")
    pg_0 = dist.new_group(ranks=[0, 1, 2, 3])
    pg_1 = dist.new_group(ranks=[0, 2, 4, 6])

    if dist.get_rank() in [0, 1, 2, 3]:
        dist.barrier(pg_0)

    # From global rank 0's perspective,
    # - dist.get_world_size(pg_0) == dist.get_world_size(pg_1) == 0
    # - dist.get_rank(pg_0) == dist.get_rank(pg_1) == 0
    #
    # However the two groups are not interchangeable. For example, the
    # following code would deadlock:
    if dist.get_rank() == 0:
        dist.barrier(pg_1)
    if dist.get_rank() in [1, 2, 3]:
        dist.barrier(pg_0)


if __name__ == "__main__":
    lc = pet.LaunchConfig(
        min_nodes=1,
        max_nodes=1,
        nproc_per_node=8,
        run_id=str(uuid.uuid4()),
        rdzv_backend="c10d",
        rdzv_endpoint="localhost:0",
        monitor_interval=1,
    )
    pet.elastic_launch(lc, entrypoint=main)()
```

@mrshenli might be able to provide info on what to guard.

wconstab replied:
well, this doesn't totally preclude the 'decaying' approach - as long as we also include another identifier in the decay.

however, as @voz pointed out we can also consider tracing the pg as a 'global' and simply guarding that it hasn't changed.  We can combine both ID_CHECK and matching of key attributes to deal with the issue of ambiguity based on attributes alone.

wconstab committed:
Update on "Dynamo traces distributed.allreduce"


- a new guard type (PROCESS_GROUP_MATCH) enforces critical properties
  of the PG match at runtime, while not exhaustively safeguarding
  other transient state vars of c++ PG impl
- allreduce is inplace-only, and stream is ignored

Based on work from Yifu Wang

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

wconstab committed:
Update on "Dynamo traces distributed.allreduce"


- a new guard type (PROCESS_GROUP_MATCH) enforces critical properties
  of the PG match at runtime, while not exhaustively safeguarding
  other transient state vars of c++ PG impl
- allreduce is inplace-only, and stream is ignored

Based on work from Yifu Wang

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]
